container-engine-lib/lib/backend_impls/docker/docker_kurtosis_backend/backend_creator/backend_creator.go

```diff
@@ -30,24 +42,123 @@ type APIContainerModeArgs struct {
+	if err != nil {
+		return nil, stacktrace.Propagate(err, "Error building Kurtosis remote Docker backend")
+	}
+	return kurtosisRemoteBackend, nil
+}
+
+func buildRemoteDockerClient(remoteBackendConfig *configs.KurtosisRemoteBackendConfig) (*client.Client, error) {
+	var clientOptions []client.Opt
+
+	// host and port option
+	clientOptions = append(clientOptions, client.WithHost(remoteBackendConfig.Endpoint))
+
+	// TLS option if config is present
+	if tlsConfig := remoteBackendConfig.Tls; tlsConfig != nil {
+		tlsFilesDir, cleanCertFilesFunc, err := writeTlsConfigToTempDir(tlsConfig.Ca, tlsConfig.ClientCert, tlsConfig.ClientKey)
+		if err != nil {
+			return nil, stacktrace.Propagate(err, "Error building TLS configuration to connect to remote Docker backend")
+		}
+		defer cleanCertFilesFunc()
+		tlsOpt := client.WithTLSClientConfig(
+			path.Join(tlsFilesDir, caFileName),
+			path.Join(tlsFilesDir, certFileName),
+			path.Join(tlsFilesDir, keyFileName))
+		clientOptions = append(clientOptions, tlsOpt)
+	}
+
+	// API version negotiation option
+	clientOptions = append(clientOptions, client.WithTimeout(dockerClientTimeout), client.WithAPIVersionNegotiation())
+
+	remoteDockerClient, err := client.NewClientWithOpts(clientOptions...)
+	if err != nil {
+		return nil, stacktrace.Propagate(err, "Error building Docker remote client")
+	}
+	return remoteDockerClient, nil
+}
+
+// writeTlsConfigToTempDir writes the different TLS files to a directory, and returns the path to this directory.
+// It also returns a function to manually delete those files once they've been used upstream
+func writeTlsConfigToTempDir(ca []byte, cert []byte, key []byte) (string, func(), error) {
+	tempDirectory, err := os.MkdirTemp(noTempDirPrefix, tempDirNamePattern)
+	if err != nil {
+		return "", nil, stacktrace.Propagate(err, "Cannot create a temporary directory to store Kurtosis backend TLS files")
+	}
+	caAbsFileName := path.Join(tempDirectory, caFileName)
+	if err = os.WriteFile(caAbsFileName, ca, tlsFilesPerm); err != nil {
+		return "", nil, stacktrace.Propagate(err, "Error writing content of CA to temporary file at '%s'", caAbsFileName)
+	}
+	certAbsFileName := path.Join(tempDirectory, certFileName)
+	if err = os.WriteFile(certAbsFileName, cert, tlsFilesPerm); err != nil {
+		return "", nil, stacktrace.Propagate(err, "Error writing content of certificate to temporary file at '%s'", certAbsFileName)
+	}
+	keyAbsFileName := path.Join(tempDirectory, keyFileName)
+	if err = os.WriteFile(keyAbsFileName, key, tlsFilesPerm); err != nil {
+		return "", nil, stacktrace.Propagate(err, "Error writing content of key to temporary file at '%s'", keyAbsFileName)
+	}
+
+	cleanDirectoryFunc := func() {
+		if err = os.RemoveAll(tempDirectory); err != nil {
+			logrus.Warnf("Error removing TLS config directory at '%s'. Will remain in the OS temporary files folder until the OS removes it", tempDirectory)
+		}
+	}
+	return tempDirectory, cleanDirectoryFunc, nil
```

Comment:
This code looks strangely familiar to me. Was it copied or moved from somewhere? If it's used by multiple things, maybe we can put it in a shared lib or something to DRY this out a bit

Reply:
I moved some of this code from the remote context backend.